csharp/src/Google.Protobuf/ParsingPrimitives.cs

```diff
@@ -638,17 +636,68 @@ public static string ReadRawString(ref ReadOnlySpan<byte> buffer, ref ParserInte
+            return ReadStringSlow(ref buffer, ref state, length);
+        }
+
+        /// <summary>
+        /// Reads a string assuming that it is spread across multiple spans in a <see cref="ReadOnlySequence{T}"/>.
+        /// </summary>
+        private static string ReadStringSlow(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state, int length)
+        {
+#if GOOGLE_PROTOBUF_SUPPORT_FAST_STRING
+            if (IsDataAvailable(ref state, length))
+            {
+                if (state.stringDecoder == null)
+                {
+                    state.stringDecoder = CodedOutputStream.Utf8Encoding.GetDecoder();
+                }
+
+                // We need to decode bytes incrementally across multiple spans.
+                int maxCharLength = CodedOutputStream.Utf8Encoding.GetMaxCharCount(length);
+                char[] charArray = ArrayPool<char>.Shared.Rent(maxCharLength);
+
+                try
+                {
+                    int remainingByteLength = length;
+                    int initializedChars = 0;
+                    while (remainingByteLength > 0)
+                    {
+                        if (buffer.Length - state.bufferPos == 0)
+                        {
+                            state.segmentedBufferHelper.RefillBuffer(ref buffer, ref state, true);
+                        }
+
+                        var unreadSpan = buffer.Slice(state.bufferPos, buffer.Length - state.bufferPos);
+                        int bytesRead = Math.Min(remainingByteLength, unreadSpan.Length);
+                        remainingByteLength -= bytesRead;
+                        bool flush = remainingByteLength == 0;
+
+                        unsafe
+                        {
+                            fixed (byte* pUnreadSpan = &MemoryMarshal.GetReference(unreadSpan))
+                            fixed (char* pCharArray = &charArray[initializedChars])
+                            {
+                                initializedChars += state.stringDecoder.GetChars(pUnreadSpan, bytesRead, pCharArray, charArray.Length - initializedChars, flush);
+                            }
+                        }
 
-            // TODO: even if GOOGLE_PROTOBUF_SUPPORT_FAST_STRING is not supported,
-            // we could still create a string efficiently by using Utf8Encoding.GetString(byte[] bytes, int index, int count)
-            // whenever the buffer is backed by a byte array (and avoid creating a new byte array), but the problem is
-            // there is no way to get the underlying byte array from a span.
+                        state.bufferPos += bytesRead;
+                    }
 
-            // TODO: in case the string spans multiple buffer segments, creating a char[] and decoding into it and then
-            // creating a string from that array might be more efficient than creating a string from the copied bytes.
+                    string value = new string(charArray, 0, initializedChars);
```

Comment:
Based on the documentation, this constructor still creates a new char array internally (which is also why you're returning the rented char array).
That also means
- you're renting a charArray (which is approx. 2*length bytes long) and then decode into the charArray with not-so-trivial code
- then you copy the decoded chars into a new array anyway.

So compared to the current approach, the only improvement seems to be the use of arrayPool (vs just allocating new byte[] array). (Benchmark numbers seem to confirm that).
What you could do is to just rent a byte[] array, read raw bytes into it (much simpler than using a Decoder) and then create a new string by decoding that byte array (and finally return the rented byte array). You'd even need to rent less memory in most cases (because you only need 1byte vs 2bytes for some common characters).

Reply:
Good point. Changed